### ISSUE TITLE
Problems with current helm configuration

### ISSUE THREAD

**Issue by Laremere** [googleforgames/open-match#1244]
Preface: I have not looked closely at how all of our helm stuff works, so some of this may be way off.

Our helm charts seem overly complicated, and yet also under powered.  I want to do a deep dive on how it's all set up at some point.  Specifically looking at these items:

- Our charts seem to have deviated from standard practice, see https://github.com/googleforgames/open-match/issues/1235
- The customize subchart should be removed.  We need a chart for the default evaluator, and also a chart for the default evaluator.  If you want a custom evaluator, or are deploying your match function, you should be deploying it like any other k8s service.  Adding our layer on top only confuses things.
- values.yaml and the OM config map should converge, or at least directly include the om config map object in the values.yaml.   This leads to overall system complexity with little point.
- In hindsight, the override configmap is proving to be confusing.  I still maybe think there's some system level stuff which could be configured in a different spot, but maybe not.
- Composition over inclusion: Currently the OM helm chart includes several other charts as sub charts.  This leads to a lot of complexity, and makes certain things more confusing than it needs to be (eg, using your own redis).  It also adds layering problems where all of the configuration you may want to have for redis needs to be pumped through the OM values.yaml.  I don't know if the end result will be simpler without trying, but I'd like to try something like "step 1: Install redis, you can use this chart.  step 2: install OM, configure to point to your redis by setting this config value."  We should provide a nice all in one yaml for the quickstart/demo, but quickly break out after that.
- Specific customization of each component's resources.
- Query should be on a vertical scaling policy, instead of horizontal.

If anyone has opinions, please share.

**Comment by TBBle** [comment]
> * We need a chart for the default evaluator, and also a chart for the default evaluator.

Was one of those "evaluator" occurrences supposed to be "match function"?

**Comment by Laremere** [comment]
I think what I meant to say was one for the demo, fixed.  I specifically don't want a "use this chart for your match function" because we don't know what requirements/configuration your match function needs, and providing a chart which does so is confusing people into thinking they need to use it.

**Comment by TBBle** [comment]
Yeah, that makes sense for open-match-customize. Really, it should be called "open-match-default-replacables" or something, to be clear that this is what it's doing.

It's a shame the evaluator must have its details listed in the open-match configuration, this makes it a little annoying to remove the open-match-customize subchart; you must then provide settings for `evaluator.hostName` etc. The matchFunction helpfully does not need to be configured at Open Match install time, since the Director provides it... Would it make sense for the evaluator to also be named by the Director?

----

I agree having a redis sub-chart is useful for quick-start, and that it should be nearly-trivial to disable it and provide your own redis connection details.

----

Another suggestion: Right now, the image tags are hard-coded in the values.yaml, and happen to be the same as the chart versions. It would be possible to have the image tags fall back to `Chart.AppVersion` and `Chart.Version`, and default `global.image.tag` to empty. However, you would need to roll the open-match-scale subchart back into the main chart, so that its image tag can have access to the main chart's Version. And the same for open-match-customize if it's not removed by then.

I'm guessing that the release process for Open Match Helm charts edits the values.yaml to set the global.image.tag to the release version? By using `Chart.*Version` for tags instead, the value would come from the value used with `helm package --version X` (or `--appVersion`, if used), simplifying releases slightly as there's no file-editing required.

**Comment by Laremere** [comment]
Multiple fetch match calls will use the same evaluator, which is why it isn't specified in the call.

The tags suggestion is interesting.  I'm always for reducing complexity.

Speaking, I'm wondering why the various deployments, services, and autoscalers don't have a template in the _helpers.yaml?  It seems like a lot of complexity could be removed, especially since most of the helpers are injected into all of the deployments.

**Comment by TBBle** [comment]
While I think of it, I want to surface the suggestion I made in https://github.com/googleforgames/open-match/issues/1235#issuecomment-665693441, that the [Standard Labels](https://helm.sh/docs/chart_best_practices/labels/) be added to all the charts. Even the older-style metadata used now (`app`, etc.) are annotations rather than labels in some cases, I found last night.

**Comment by Laremere** [comment]
There's some problems with the config being generated as well:

- Resources for the different services should be more independently controllable, as eg the synchronizer commonly has higher load and can't be scaled, while the frontend is better scaled with more instances.
- Query should default to 3 instances with vertical autoscaling instead of horizontal, to better handle how it needs to scale: It builds memory pressure, and horizontal auto scaling doesn't relieve it, so it scales to the max instances pointlessly.  Also fewer query instances means less load on redis.

**Comment by TBBle** [comment]
I also don't understand what the open-match-scale subchart is for. Those appear to be part of a scaling demo, I feel like the demo should be depending on (subcharting) Open Match, not the other way 'round.

**Comment by Laremere** [comment]
Yeah, that's a pretty obvious change now that you mention it.  We should probably do the same with the standard demo as well.
